LUN-274: Select profession on initial start

SteffiStuffel opened the pull request:
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

klinzo commented on src/navigation/Navigator.tsx:
```diff
@@ -56,13 +60,25 @@ const styles = (headerHeight?: number) =>
   // Set only height for tablets since header doesn't scale auto
   const headerHeight = useTabletHeaderHeight(wp('15%'))
 
+  useEffect(() => {
+    const initialize = async () => {
+      const professions = await AsyncStorage.getSelectedProfessions()
+      setIntroShown(professions !== null)
+    }
+    initialize()
+      .then(() => setWaitingForInit(false))
+      .catch(error => reportError(error))
+  }, [])
```
Why do we need a custom useEffect here instead of just using `useReadSelectedProfession`? Then you can just do `const introShown = professions !== null`

SteffiStuffel replied:
I see your point here, but not sure if it is possible to do. Profession is null, either when a) intro was not shown before or b) when profession is not yet loaded (when using useReadSelectedProfession). Therefore we need to check for the Promise to return, to set waitingForInti(false) correctly.

klinzo replied:
You can just use `loading` property returned by the hook instead of `waitingForInit` or am I missing something?

SteffiStuffel replied:
Nope, you are right, this is working.

klinzo commented on src/navigation/Navigator.tsx:
```diff
@@ -56,13 +60,25 @@ const styles = (headerHeight?: number) =>
+  useEffect(() => {
+    const initialize = async () => {
+      const professions = await AsyncStorage.getSelectedProfessions()
+      setIntroShown(professions !== null)
```
Is it required that a profession gets selected? Maybe it would be more explicit if we saved a separate value `introShown` to the async storage.

I also think that if I don't want to select a profession I should not have to click on `Überspringen` each time I open the app (this fails the alternate course of the issue).

SteffiStuffel replied:
Yes, currently profession is null, when the value was never set. If you skip the selection, an empty array should be saved, therefore the alternate course should work. But I forgot to set this. 

If you think the clearness of a separate value will outweigh that one more slow reading from AsyncStorage we need then, I will add it. Let me know what you prefer.

klinzo replied:
As discussed the current way is fine.

klinzo commented on src/routes/intro/IntroScreen.tsx:
```diff
@@ -0,0 +1,83 @@
+  const { data: selectedProfessions, refresh: refreshSelectedProfessions } = useReadSelectedProfessions()
+
+  useFocusEffect(
+    React.useCallback(() => {
+      refreshSelectedProfessions()
+    }, [refreshSelectedProfessions])
```
The wrapping in callback should not be needed as it is already wrapped in a callback in `useLoadAsync`

klinzo commented on src/routes/intro/IntroScreen.tsx:
```diff
@@ -0,0 +1,83 @@
+
+const IntroScreen = ({ navigation }: IntroScreenProps): JSX.Element => {
+  const { data: disciplines, error, loading, refresh } = useLoadDisciplines(null)
+  const { data: selectedProfessions, refresh: refreshSelectedProfessions } = useReadSelectedProfessions()
```
I think we have some inconsistent naming again now. Profession vs Discipline. Don't we use that the same way?

SteffiStuffel replied:
Yes there is a reason: professions are disciplines that can be selected (it is only possible for second level disciplines of lunes standard vocabulary)

klinzo commented on src/routes/intro/IntroScreen.tsx:
```diff
@@ -0,0 +1,83 @@
+
+const IntroScreen = ({ navigation }: IntroScreenProps): JSX.Element => {
+  const { data: disciplines, error, loading, refresh } = useLoadDisciplines(null)
+  const { data: selectedProfessions, refresh: refreshSelectedProfessions } = useReadSelectedProfessions()
```
Why don't you handle error and loading the same way as those from `useLoadDiscipline`?

SteffiStuffel replied:
If `useLoadDiscipline` is still loading, we have no data to display, whereas if the `useReadSelectedProfession` is still loading, we can display the data, but only the green-check-icon is missing. Not sure if the reading from AsyncStorage can fail :thinking: At least it is far less common than a failing api-request.

If you want I can open a ticket, where we implement an appropriate error handling.

klinzo replied:
No, sounds good to me :+1:

klinzo commented on src/services/AsyncStorage.ts:
```diff
@@ -1,5 +1,41 @@
+  await AsyncStorage.setItem('selectedProfessions', JSON.stringify(selectedProfessions))
+}
+
+export const pushSelectedProfession = async (profession: Discipline): Promise<Discipline[]> => {
```
Some unit tests would be nice here

SteffiStuffel replied:
done.

klinzo commented on src/routes/intro/__tests__/IntroScreen.spec.tsx:
```diff
@@ -0,0 +1,55 @@
+describe('IntroScreen', () => {
+  const navigation = createNavigationMock<'Intro'>()
+
+  it('should navigate to discipline', async () => {
```
```suggestion
  it('should navigate to profession selection of discipline', async () => {
```

klinzo commented on src/routes/intro/__tests__/IntroScreen.spec.tsx:
```diff
@@ -0,0 +1,55 @@
+    mocked(useReadSelectedProfessions).mockReturnValueOnce(getReturnOf(null))
+
+    const { findByText } = render(<IntroScreen navigation={navigation} />)
+    const firstDiscipline = await findByText('First Discipline')
```
No need for `await findByText`, `getByText` is enough

SteffiStuffel replied:
Why? Because I mocked every async stuff and it is no longer asnyc then?

klinzo replied:
yes, exactly.

klinzo commented on src/routes/intro/__tests__/ProfessionSelectionScreen.spec.tsx:
```diff
@@ -0,0 +1,73 @@
+    expect(profession).toBeDefined()
+    expect(queryAllByTestId('check-icon')).toHaveLength(0)
+    fireEvent.press(profession)
+    expect(spyOnPush).toHaveBeenCalledTimes(1)
+    expect(spyOnPush).toHaveBeenCalledWith(mockDisciplines[0])
```
Optional to avoid spies:
```suggestion
    await waitFor(async () => {
      const selectedProfessions = await AsyncStorageService.getSelectedProfessions()
      expect(selectedProfessions).toEqual([mockDisciplines[0]])
    })

```

SteffiStuffel replied:
Done.
I decided to use a spy, as I thought it makes sense for a unit test, to just check if "outgoing" call is valid and not also if the AsyncStorageService is working properly. Why is it bad to use spies?
Best test case, would be to wait for green-check-icon to appear / disapear, but I cannot get this working.

klinzo replied:
I don't know if it is bad to use spies but I think it kind of tests implementation details which could be avoided here checking if the async storage has the right value afterwards. E.g. it could happen that while the spy is called correctly, another async storage function is called somewhere else overwriting it in the async storage.

klinzo commented on src/routes/intro/ProfessionSelectionScreen.tsx:
```diff
@@ -0,0 +1,121 @@
+          contentContainerStyle={{ flexGrow: 1 }}
+          ListHeaderComponent={<Title title={discipline.title} description={childrenDescription(discipline)} />}
+          ListFooterComponent={
+            !selectedProfessions || selectedProfessions.length === 0 ? (
+              <ButtonContainer>
+                <Button
+                  onPress={navigateToHomeScreen}
+                  label={labels.intro.skipSelection}
+                  buttonTheme={BUTTONS_THEME.contained}
+                />
+              </ButtonContainer>
+            ) : (
+              <></>
+            )
```
I think it would be more intuitive (and easier to implement) if the skip button would be implemented the same way as the confirm selection button. Otherwise the skip button is mostly not visible and the user has to scroll down first. In combination with a very unobtrusive/small text that the user should select their professions I think it is kind of hard to grasp what the user has to do.

SteffiStuffel replied:
It was on purpose the "skip" button is kind of "hidden", so the users get encouraged to select a profession. Would keep it for now like this.

klinzo replied:
Okay, then LGTM.

SteffiStuffel commented on src/hooks/useReadSelectedProfessions.ts:
```diff
@@ -0,0 +1,9 @@
+import AsyncStorage from '../services/AsyncStorage'
+import useLoadAsync, { Return } from './useLoadAsync'
+
+const loadFromAsyncStorage = async (): Promise<Discipline[] | null> => AsyncStorage.getSelectedProfessions()
```
Done, also for useReadCustomDisicplines.

SteffiStuffel commented on src/hooks/useReadSelectedProfessions.ts:
```diff
@@ -0,0 +1,9 @@
+
+const loadFromAsyncStorage = async (): Promise<Discipline[] | null> => AsyncStorage.getSelectedProfessions()
+
+const useReadFromAsyncStorage = (): Return<Discipline[] | null> => useLoadAsync(loadFromAsyncStorage, null)
```
I think as it is/ can be used in different parts of the code it is good to have it in a separate file.

SteffiStuffel commented on src/routes/intro/ProfessionSelectionScreen.tsx:
```diff
@@ -0,0 +1,121 @@
+  height: 100%;
+`
+
+const StyledList = styled(FlatList)`
+  width: 100%;
+` as ComponentType as new () => FlatList<Discipline>
```
hmh, I think you are right, it is not needed. :thinking:

SteffiStuffel commented on src/routes/intro/ProfessionSelectionScreen.tsx:
```diff
@@ -0,0 +1,121 @@
+      rightChildren={
+        selectedProfessions?.includes(item) ? (
+          <IconContainer>
+            <CheckCircleIconGreen testID='check-icon' />
```
Fixed it.

f1sh1918 reviewed (approved):
Tested on ios. Button "Starten" could need some margin-bottom to avoid accidently clicking on the ios bottom sheet. See suggestion (optional)
![image](https://user-images.githubusercontent.com/37902063/161717415-53efa626-0c6e-462b-9069-56ffc6687741.png)

f1sh1918 commented on src/routes/intro/ProfessionSelectionScreen.tsx:
```diff
@@ -0,0 +1,121 @@
+`
+
+const ButtonContainer = styled.View`
+  padding-top: ${props => props.theme.spacings.md}
```
```suggestion
  padding: ${props => `${props.theme.spacings.md} 0`}
```

SteffiStuffel replied:
done

f1sh1918 commented on src/routes/intro/ProfessionSelectionScreen.tsx:
```diff
@@ -0,0 +1,121 @@
+                />
+              </ButtonContainer>
+            ) : (
+              <></>
```
no need to render a fragment if there is no content in it. Instead return `: null`

klinzo commented on src/routes/intro/ProfessionSelectionScreen.tsx:
```diff
@@ -48,9 +44,10 @@ const ProfessionSelectionScreen = ({ route, navigation }: ProfessionSelectionScr
   const { discipline } = route.params
   const { data: disciplines, error, loading, refresh } = useLoadDisciplines(discipline)
   const { data: selectedProfessions, refresh: refreshSelectedProfessions } = useReadSelectedProfessions()
+  const isProfessionSelected = selectedProfessions && selectedProfessions.length > 0
```
Optional: Imo naming should be plural, otherwise I'd think it is just whether a currently displayed profession is selected and not whether any profession is selected.

klinzo commented on src/routes/intro/ProfessionSelectionScreen.tsx:
```diff
@@ -74,19 +71,22 @@ const ProfessionSelectionScreen = ({ route, navigation }: ProfessionSelectionScr
 
-  const navigateToHomeScreen = () => {
+  const navigateToHomeScreen = async () => {
+    if (!isProfessionSelected) {
```
Same as in IntroScreen

klinzo commented on src/routes/intro/IntroScreen.tsx:
```diff
@@ -37,21 +38,21 @@ interface IntroScreenProps {
-  const navigateToHomeScreen = (): void => {
-    navigation.navigate('Home')
+  const navigateToHomeScreen = async () => {
+    if (!isProfessionSelected) {
```
I'd just check `selectedProfessions === null`, no need to set the selected professions to `[]` if it is already `[]`

SteffiStuffel replied:
you are right. done

klinzo commented on src/routes/intro/IntroScreen.tsx:
```diff
@@ -37,21 +38,21 @@ interface IntroScreenProps {
+    if (!isProfessionSelected) {
+      await AsyncStorage.setSelectedProfessions([])
+    }
+    navigation.push('Home')
```
Why push instead of navigate? Same for ProfessionSelectionScreen

SteffiStuffel replied:
Changed to navigate

klinzo reviewed (commented):
There are still some comments that are hidden and where not addressed

klinzo reviewed (approved):
Looking good. Please have a look at my comments, you don't have to do everything (some hidden comments not yet addressed from my first review).

SteffiStuffel commented:
> There are still some comments that are hidden and are not addressed yet (maybe also on purpose, not sure)

I think I addressed all comments. Also fixed the bottom margin @f1sh1918 mentioned.
